### ISSUE TITLE
The number of selected features == population size?

### ISSUE THREAD

**Issue by imxtx** [rodrigo-arenas/Sklearn-genetic-opt#141]
After feature selection, I transform the training data, but the number of the selected features is not what I want (100):

Feature selection:
```
evolved_estimator = GAFeatureSelectionCV(
        estimator=pipline,
        scoring="accuracy",
        population_size=100,
        generations=200,
        mutation_probability=mutation_scheduler,
        crossover_probability=crossover_scheduler,
        n_jobs=-1,
    )
```

After selection, transform the training data:
```
X = estimator.transform(X)
print(X[0].shape) # print 607, should be 100
```

What should I do if I only want the most important 100 features? I already saved the estimator, can I get the most important 100 features from the dumped estimator?

**Comment by rodrigo-arenas** [comment]
Hi @imxtx 
The population_size controls how many different individuals (models) are created at generation zero, if you want to control the number of features, you can use the parameter `max_features` and set it to 100 or so.
Keep in mind that the package will try to not choose more than 100 features in that case, but there is no guarantee that it will select exactly 100 features

So the code change to something like this:

```python
evolved_estimator = GAFeatureSelectionCV(
        estimator=pipline,
        scoring="accuracy",
        population_size=30,
        generations=25,
        max_features=100,
        mutation_probability=mutation_scheduler,
        crossover_probability=crossover_scheduler,
        n_jobs=-1,
    )
```

**Comment by imxtx** [comment]
does the estimator choose the features by their importance ranks？is it possible to get an importance rank for each feature? 
    
        
        
            
                
                    Tianxin Xie
                    
                    
                                
                                    ***@***.***
                                
                
            
        
        
    

---- Replied Message ----FromRodrigo ***@***.***>Date09/19/2023 22:51 ***@***.***> CcT.X. ***@***.***>***@***.***>SubjectRe: [rodrigo-arenas/Sklearn-genetic-opt] The number of selected features == population size? (Issue #141)
Hi @imxtx
The population_size controls how many different individuals (models) are created at generation zero, if you want to control the number of features, you can use the parameter max_features and set it to 100 or so.
Keep in mind that the package will try to not choose more than 100 features in that case, but there is no guarantee that it will select exactly 100 features
So the code change to something like this:
evolved_estimator = GAFeatureSelectionCV(
        estimator=pipline,
        scoring="accuracy",
        population_size=30,
        generations=25,
        max_features=100,
        mutation_probability=mutation_scheduler,
        crossover_probability=crossover_scheduler,
        n_jobs=-1,
    )

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>
[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/rodrigo-arenas/Sklearn-genetic-opt/issues/141#issuecomment-1725800918",
"url": "https://github.com/rodrigo-arenas/Sklearn-genetic-opt/issues/141#issuecomment-1725800918",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by rodrigo-arenas** [comment]
It doesn't quite work that way, it will create individuals who represent a set of possible features, then the model is fitted and the whole set of features and is ranked using the cross-validation scores, so there is not individual importance of each feature, but the subset of features as a whole.
You can explore the `evolved_estimator.cv_results_`  and the `evolved_estimator.logbook` to check the history of all the estimators

**Comment by imxtx** [comment]
I see. thank you very much！ 
    
        
        
            
                
                    Tianxin Xie
                    
                    
                                
                                    ***@***.***
                                
                
            
        
        
    

---- Replied Message ----FromRodrigo ***@***.***>Date09/19/2023 23:07 ***@***.***> CcT.X. ***@***.***>***@***.***>SubjectRe: [rodrigo-arenas/Sklearn-genetic-opt] The number of selected features == population size? (Issue #141)
It doesn't quite work that way, it will create individuals who represent a set of possible features, then the model is fitted and the whole set of features and is ranked using the cross-validation scores, so there is not individual importance of each feature, but the subset of features as a whole.
You can explore the evolved_estimator.cv_results_  and the evolved_estimator.logbook to check the history of all the estimators

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>
[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/rodrigo-arenas/Sklearn-genetic-opt/issues/141#issuecomment-1725836128",
"url": "https://github.com/rodrigo-arenas/Sklearn-genetic-opt/issues/141#issuecomment-1725836128",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by rodrigo-arenas** [comment]
I'm closing this issue, let me know if you have more questions
